src/reporters/json.ts

```diff
@@ -27,18 +27,119 @@ import BaseReporter from './base';
 import { formatError, getTimestamp } from '../helpers';
 import { Journey, Step } from '../dsl';
 import snakeCaseKeys from 'snakecase-keys';
+import { NetworkInfo } from '../common_types';
 
 /* eslint-disable @typescript-eslint/no-var-requires */
-const programVersion = require('../../package.json').version;
+const { version, name } = require('../../package.json');
+
+type OutputType =
+  | 'journey/register'
+  | 'journey/start'
+  | 'step/screenshot'
+  | 'step/end'
+  | 'journey/network_info'
+  | 'journey/filmstrips'
+  | 'journey/browserconsole'
+  | 'journey/end';
+
+type OutputFields = {
+  type: OutputType;
+  journey: Journey;
+  timestamp?: number;
+  url?: string;
+  step?: Partial<Step>;
+  error?: Error;
+  root_fields?: Record<string, unknown>;
+  payload?: Record<string, unknown>;
+  blob?: string;
+};
+
+function getMetadata() {
+  return {
+    process: {
+      pid: process.pid,
+      ppid: process.ppid,
+      title: process.title,
+      args: process.argv,
+    },
+    os: {
+      platform: process.platform,
+    },
+    package: {
+      name,
+      version,
+    },
+  };
+}
+
+function formatHTTPVersion(protocol: string) {
+  if (protocol === 'h2') {
+    return 2;
+  } else if (protocol === 'http/1.1') {
+    return 1.1;
+  } else if (protocol === 'http/1.0') {
+    return 1.0;
+  } else if (protocol && protocol.startsWith('h3')) {
+    return 3;
+  }
+  return;
+}
+
+export function formatNetworkFields(network: NetworkInfo) {
+  const { request, response, url } = network;
+  const postdata = request.postData || '';
+  const tls = response?.securityDetails;
+
+  return {
+    // URL and USER AGENT would be parsed and mapped by heartbeat
+    url,
+    user_agent: request.headers?.['User-Agent'],
+    http: {
+      version: formatHTTPVersion(response?.protocol),
+      request: {
+        body: {
+          bytes: postdata.length,
+          content: postdata,
+        },
+        method: request.method,
+        referrer: request.headers?.Referer,
+      },
+      response: response
```

Comment:
I notice a common pattern here of `value ? {...fields} : undefined`. It'd be better to have the key not be present if there are no fields since ES stores the full JSON source.

Reply:
when `undefined` gets `JSON.stringify(undefined)`, we wont see them on the output. So we shouldnt need to worry about them.